In [8]:
import pandas as pd
import statsmodels.api as sm
import pickle

# Step 1: Create the dataset
data = {
    'Y_obs': [137, 118, 124, 124, 120, 129, 122, 142, 128, 114,
              132, 130, 130, 112, 132, 117, 134, 132, 121, 128],
    'W':     [0, 1, 1, 1, 0, 1, 1, 0, 0, 1,
              1, 0, 0, 1, 0, 1, 0, 0, 1, 1],
    'X':     [19.8, 23.4, 27.7, 24.6, 21.5, 25.1, 22.4, 29.3, 20.8, 20.2,
              27.3, 24.5, 22.9, 18.4, 24.2, 21.0, 25.9, 23.2, 21.6, 22.8]
}
df = pd.DataFrame(data)

# Step 2: Prepare the input features and add the constant term (intercept)
X_reg = sm.add_constant(df[['W', 'X']])
Y_reg = df['Y_obs']

# Step 3: Fit the OLS regression model
model = sm.OLS(Y_reg, X_reg).fit()

# Step 4: Print the summary (optional)
print(model.summary())

# Step 5: Save the trained model to a .pkl file
with open("model.pkl", "wb") as f:
    pickle.dump(model, f)


                            OLS Regression Results                            
Dep. Variable:                  Y_obs   R-squared:                       0.698
Model:                            OLS   Adj. R-squared:                  0.662
Method:                 Least Squares   F-statistic:                     19.61
Date:                Sun, 20 Apr 2025   Prob (F-statistic):           3.84e-05
Time:                        17:14:01   Log-Likelihood:                -57.076
No. Observations:                  20   AIC:                             120.2
Df Residuals:                      17   BIC:                             123.1
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         95.9662      8.919     10.760      0.0

Parameter | Estimate | Std. Error | t-value | p-value | 95% CI

α (const) | 95.97 | 8.92 | 10.76 | 0.000 | [77.15, 114.78]
τ (W) | –9.11 | 2.05 | –4.43 | 0.000 | [–13.44, –4.77]
β (X) | 1.51 | 0.37 | 4.06 | 0.001 | [0.73, 2.30]